dbcon/joblist/groupconcat.cpp

```diff
@@ -758,60 +1023,196 @@ uint64_t GroupConcatOrderBy::getKeyLength() const
+  fDataVec.resize(sz);
+  if (sz > 0)
+  {
+    for (uint64_t i = 0; i < sz; ++i)
+    {
+      fDataVec[i].reset(new rowgroup::RGData(fRowGroup, fRowsPerRG));
+      fDataVec[i]->deserialize(bs, fRowGroup.getDataSize(fRowsPerRG));
+      fRowGroup.setData(fDataVec[i].get());
+      auto rgsize = fRowGroup.getSizeWithStrings(fRowsPerRG);
+      fMemSize += rgsize;
+    }
+
+    fRowGroup.initRow(&fRow0);
+    fRowGroup.getRow(fRowGroup.getRowCount() - 1, &fRow0);
+  }
+  else
+  {
+    createNewRGData();
+  }
+
+  fRule.fIdbCompare = this;
+  for (auto& compare : fRule.fCompares)
+  {
+    delete compare;
+    compare = nullptr;
+  }
+  fRule.fCompares.clear();
+  fRule.compileRules(fOrderByCond, fRowGroup);
+  fRowGroup.initRow(&row1);
+  fRowGroup.initRow(&row2);
+
+  bs >> tmp8;
+  fDistinct = tmp8;
+  if (fDistinct)
+  {
+    bs >> sz;
+    fDistinctMap.reset(new DistinctMap(sz, Hasher(this, getKeyLength()), Eq(this, getKeyLength())));
+    for (uint64_t i = 0; i < sz; ++i)
+    {
+      uint64_t idx;
+      bs >> idx;
+      uint64_t gid = idx >> 16;
+      uint64_t rid = idx & 0xffff;
+      rowgroup::Row row;
+      fRowGroup.setData(fDataVec[gid].get());
+      fRowGroup.initRow(&row);
+      fRowGroup.getRow(rid, &row);
+      fDistinctMap->emplace(row.getPointer(), idx);
+    }
+  }
+
+  bs >> sz;
+  fOrderByQueue.reset(new SortingPQ(sz));
+  for (uint64_t i = 0; i < sz; ++i)
+  {
+    uint64_t idx;
+    bs >> idx;
+    uint64_t gid = idx >> 16;
+    uint64_t rid = idx & 0xffff;
+    rowgroup::Row row;
+    fRowGroup.setData(fDataVec[gid].get());
+    fRowGroup.initRow(&row);
+    fRowGroup.getRow(rid, &row);
+    fOrderByQueue->push(GroupConcatOrderByRow(row, idx, fRule));
+  }
+  fRowGroup.setData(fDataVec.back().get());
+  fRowGroup.getRow(fRowGroup.getRowCount() - 1, &fRow0);
+}
+
+void GroupConcatOrderBy::createNewRGData()
+{
+  auto newSize = fRowGroup.getSizeWithStrings(fRowsPerRG);
+
+  fMemSize += newSize;
+
+  fDataVec.emplace_back(make_unique<rowgroup::RGData>(fRowGroup, fRowsPerRG));
+  fRowGroup.setData(fDataVec.back().get());
+  fRowGroup.setUseOnlyLongString(true);
+  fRowGroup.resetRowGroup(0);
+  fRowGroup.initRow(&fRow0);
+  fRowGroup.getRow(0, &fRow0);
+}
+
+rowgroup::RGDataSizeType GroupConcatOrderBy::getDataSize() const
+{
+  return fMemSize + fOrderByQueue->capacity() * sizeof(GroupConcatOrderByRow) +
+         (fDistinct ? fDistinctMap->size() : 0) * 32 /* TODO: speculative unordered_map memory consumption per
+                                                        item, replace it with counting allocator */
+      ;
+}
+
 void GroupConcatOrderBy::processRow(const rowgroup::Row& row)
 {
   // check if this is a distinct row
   if (fDistinct && fDistinctMap->find(row.getPointer()) != fDistinctMap->end())
     return;
 
   // this row is skipped if any concatenated column is null.
-  if (concatColIsNull(row))
+  if (!fIsJsonArrayAgg && concatColIsNull(row))
     return;
 
-  auto& orderByQueue = getQueue();
```

Comment:
Could you elaborate on this? I think this and the corresponding changes down the code should be reversed.